cherrypy/lib/covercp.py

```diff
@@ -334,26 +334,26 @@ def menu(self, base='/', pct='50', showpct='',
         yield '</body></html>'
 
     def annotated_file(self, filename, statements, excluded, missing):
-        source = open(filename, 'r')
-        buffer = []
-        for lineno, line in enumerate(source.readlines()):
-            lineno += 1
-            line = line.strip('\n\r')
-            empty_the_buffer = True
-            if lineno in excluded:
-                template = TEMPLATE_LOC_EXCLUDED
-            elif lineno in missing:
-                template = TEMPLATE_LOC_NOT_COVERED
-            elif lineno in statements:
-                template = TEMPLATE_LOC_COVERED
-            else:
-                empty_the_buffer = False
-                buffer.append((lineno, line))
-            if empty_the_buffer:
-                for lno, pastline in buffer:
-                    yield template % (lno, cgi.escape(pastline))
-                buffer = []
-                yield template % (lineno, cgi.escape(line))
+        with open(filename, 'r') as source:
+            buffer = []
+            for lineno, line in enumerate(source.readlines()):
+                lineno += 1
+                line = line.strip('\n\r')
+                empty_the_buffer = True
+                if lineno in excluded:
+                    template = TEMPLATE_LOC_EXCLUDED
+                elif lineno in missing:
+                    template = TEMPLATE_LOC_NOT_COVERED
+                elif lineno in statements:
+                    template = TEMPLATE_LOC_COVERED
+                else:
+                    empty_the_buffer = False
+                    buffer.append((lineno, line))
+                if empty_the_buffer:
+                    for lno, pastline in buffer:
+                        yield template % (lno, cgi.escape(pastline))
+                    buffer = []
+                    yield template % (lineno, cgi.escape(line))
```

Comment:
because there's a `yield` inside this `with ` you're relying on the GC to `.close()` this generator to close the file on `GeneratorExit`

I think you can fix it by using:

```python
yield from self.annotated_file(filename, statements, excluded,
                                        missing):
```

in `def report(self, name):`

Reply:
Ok, I've changed that. Alternatively, I suppose we could just put the list from `readlines()` into a variable and have it closed before the iteration.

Reply:
:facepalm: oh I only just noticed the `readlines()` call.

```python
with open(filename, 'r') as f:
    lines = f.readlines()
```

would be much more clear.

however maybe that's an error and the file was intended to be streamed? `for lineno, line in enumerate(source):`

Reply:
Originally the file was streamed in this generator and then it was changed to `.readlines()` https://github.com/cherrypy/cherrypy/commit/596b1f603047a407bf8949b018671c7de13e9f2f#diff-342872edf8db6cab0554b70351960b4850b8eb7120003bad9e2d074c69a7c758L335-R333

Reply:
Ok, I'll do the first snippet but I suppose there's no harm in leaving the `yield from` change as well.